Doc: AI Skill - Update Element Benchmarks (#1428)
BLAST-ImpactX/impactx: .claude/skills/impactx-update-element-benchmarks/SKILL.md

---
name: impactx-update-element-benchmarks
description: This skill should be used when the user asks to "update element benchmarks", "add missing element tests", "audit element test coverage", "add a benchmark for <element>", or to check that every lattice element under `src/elements/` has matching entries in the three Python element-coverage tests (benchmark, serialization, reversibility) and that spin-capable elements carry the spin/nospin parametrization. Also covers re-running those three tests after edits.
version: 0.1.0
---

# Update Element Benchmarks

Audit and extend the three Python element-coverage tests so that every lattice element exposed from `src/elements/` is represented, and that every spin-capable element is parametrized over spin tracking. Then run the three tests.

To run this skill, invoke `/impactx-update-element-benchmarks` in a new session.

> The skill also auto-loads on triggers like "update element benchmarks" or "audit element test coverage".

Reviewer: Isn't it better to include here all the examples of triggers that you listed in `description` above?

```suggestion
The skill also auto-loads on triggers like "update element benchmarks", "add missing element tests", "audit element test coverage", or "add a benchmark for <element>".
```

Author: Yes good to list them all, thanks!

**test_reversibility_elements.py** — place the new test in the correct section:

- "Thick elements (negate ds)" for anything with `ds`.
- "Thin kick elements (negate strength)" for thin-kick elements without `ds`.
- "Rotation elements" for pure rotations.
- "RF / energy-changing elements" for RF cavities.
- "Edge elements" for edge kicks (`DipEdge`, `QuadEdge`).

Use `PIPE_KWARGS` for elements that accept `aperture_x`/`aperture_y`, otherwise `ALIGNMENT_KWARGS`. Wrap the call in `roundtrip(el, sim, ...)`. Only relax `phase_atol` when a physically motivated reason exists (multi-step map, non-symplectic split, energy change); copy the rationale comment style already present.

## Pass 2 — spin parametrization audit

> An element is spin-capable iff its header under `src/elements/<Name>.H` inherits `public mixin::SpinTransport`. To find the current set in one shot:

Reviewer: Is this `iff` intentional (i.e., "if and only if") or is it a typo?

Author: haha yeah I learned about `iff` meaning in English just last week. Claude seems to like it.

> ```bash

Author: ```suggestion
```
```

# via Grep tool
Grep: "mixin::SpinTransport" under src/elements/ (files_with_matches)

> ```

Reviewer: This reads more like a prose instruction than a shell command: shouldn't it be moved out of the ` ```bash` code block?

Author: yeah or without the bash

The serialization file (file 2) is not parametrized over spin — it only checks construction and dict round-trip. Do not add spin parametrize marks there.

Elements that are NOT spin-capable (no `SpinTransport` inheritance) must not be parametrized — the parametrization will still "work" but it just runs the same `nospin` path twice, wasting CI time.

## Pass 3 — run the tests

After editing the Python test files, install the freshly-built Python bindings into the active environment and run the three test files directly with `pytest`.

```bash
# 1. Make sure a suitable micromamba env is active. Preference order:
#    1. impactx-cpu-mpich-dev   (project-specific, preferred)
#    2. warpx-cpu-mpich-dev     (BLAST developer's global default, works as fallback)
#    Do not re-activate if one is already active.

# 2. Install/refresh the Python package from the build tree.
cmake -S . -B build_cpu -DImpactX_PYTHON=ON -DImpactX_PYTHON_IPO=OFF -DpyAMReX_IPO=OFF
cmake --build build_cpu -j 6 --target pip_install

# 3. Run the three tests directly with pytest.
python -m pytest tests/pythontest_benchmark_elements.py
python -m pytest tests/pythontest_element_serialization.py

> python -m pytest tests/pythontest_reversibility_elements.py

Reviewer: I think `/` is missing from these paths:

```suggestion
# 3. Run the three tests directly with pytest.
python -m pytest tests/python/test_benchmark_elements.py
python -m pytest tests/python/test_element_serialization.py
python -m pytest tests/python/test_reversibility_elements.py
```
